webrtc-c/canary/src/Peer.cpp

```diff
@@ -487,29 +496,42 @@ STATUS Peer::handleSignalingMsg(PReceivedSignalingMessage pMsg)
 
 STATUS Peer::addTransceiver(RtcMediaStreamTrack& track)
 {
-    auto handleFrame = [](UINT64 customData, PFrame pFrame) -> VOID {
-        UNUSED_PARAM(customData);
-        // TODO: Probably reexpose or add metrics here directly
-        DLOGV("Frame received. TrackId: %" PRIu64 ", Size: %u, Flags %u", pFrame->trackId, pFrame->size, pFrame->flags);
-    };
-
     auto handleBandwidthEstimation = [](UINT64 customData, DOUBLE maxiumBitrate) -> VOID {
         UNUSED_PARAM(customData);
         // TODO: Probably reexpose or add metrics here directly
         DLOGV("received bitrate suggestion: %f", maxiumBitrate);
     };
 
+    auto handleVideoFrame = [](UINT64 customData, PFrame pFrame) -> VOID {
+        // TODO: There are times when the timestamp is corrupted in pFrame->frameData that causes very high
+        // latency metrics. Need to root cause the issue
+        PPeer pPeer = (Canary::PPeer)(customData);
+        std::unique_lock<std::recursive_mutex> lock(pPeer->mutex);
+        PBYTE pCurPtr = pFrame->frameData + pFrame->size - (UINT32) CANARY_METADATA_SIZE;
+        UINT32 size = getUnalignedInt32BigEndian((PINT32)(pCurPtr));
+        pCurPtr += SIZEOF(UINT32);
+        UINT64 tsRx = getUnalignedInt64BigEndian((PINT64)(pCurPtr));
+        pCurPtr += SIZEOF(UINT64);
+        UINT32 crc = getUnalignedInt32BigEndian((PINT32)(pCurPtr));
+        pPeer->endToEndMetricsContext.dataMatch.push_back(crc == COMPUTE_CRC32(pFrame->frameData, pFrame->size) ? 1.0 : 0.0);
```

Comment:
I think EndToEndMetricsContext should have its own lock, so that we Peer's lock is acquired the media thread doesn't block. I'm fine either way, but I feel like mutexes should be used for the smallest scope possible so that we can get the most benefit from parallelism.

Reply:
I think this should be ok. Because the callback shouldnt take long to execute.

Reply:
Fair point